src/Microsoft.Extensions.WebSockets/WebSocketConnection.cs

```diff
@@ -0,0 +1,391 @@
+        /// <param name="message">The message to send.</param>
+        /// <param name="cancellationToken">A <see cref="CancellationToken"/> that indicates when/if the send is cancelled.</param>
+        /// <returns>A <see cref="Task"/> that completes when the message has been written to the outbound stream.</returns>
+        // TODO: De-taskify this to allow consumers to create their own awaiter.
+        public Task SendAsync(WebSocketFrame message, CancellationToken cancellationToken)
+        {
+            // Calculate the size of the frame
+            // Base header size is 2 bytes.
+            var allocSize = 2;
+            var payloadLength = message.Payload.Length;
+            if (payloadLength > ushort.MaxValue)
+            {
+                // We're going to need an 8-byte length
+                allocSize += 8;
+            }
+            else if (payloadLength > 125)
+            {
+                // We're going to need a 2-byte length
+                allocSize += 2;
+            }
+            if (_maskingKey != null)
+            {
+                // We need to alloc for the whole payload (and the masking key)
+                allocSize += 4 + payloadLength;
+            }
+
+            // Allocate a buffer
+            var buffer = _outbound.Alloc(minimumSize: allocSize);
+            if (buffer.Memory.Length < allocSize)
+            {
+                throw new InvalidOperationException("Couldn't allocate enough data from the channel to write the header");
+            }
+
+            // Write the opcode and FIN flag
+            var opcodeByte = (byte)((int)message.Opcode << 4);
+            if (message.EndOfMessage)
+            {
+                opcodeByte |= 1;
+            }
+            buffer.WriteBigEndian(opcodeByte);
+
+            // Write the length and mask flag
+            var maskingByte = _maskingKey != null ? 0x01 : 0x00; // TODO: Masking flag goes here
+
+            if (payloadLength > ushort.MaxValue)
+            {
+                buffer.WriteBigEndian((byte)(0xFE | maskingByte));
+
+                // 8-byte length
+                buffer.WriteBigEndian((ulong)payloadLength);
+            }
+            else if (payloadLength > 125)
+            {
+                buffer.WriteBigEndian((byte)(0xFC | maskingByte));
+
+                // 2-byte length
+                buffer.WriteBigEndian((ushort)payloadLength);
+            }
+            else
+            {
+                // 1-byte length
+                buffer.WriteBigEndian((byte)((payloadLength << 1) | maskingByte));
+            }
+
+            // TODO: We could commit the frame header now and get the reader moving on transmitting it...
+
+            if (_maskingKey != null)
+            {
+                // TODO: Can we mask in-place in the ReadableBuffer and then append it?
+                // TODO: Vectorize
+
+                // Get a span of the output buffer for the masking key, write it there, then advance the write head.
```

Comment:
normally, you would only touch `.Memory` immediately after a `.Ensure`. Hmmm... I wonder whether we should actually make that obvious by removing `.Memory` and having it as the return value from `Ensure`